logd/server/collector.go

```diff
@@ -5,232 +5,92 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"net/http"
 	"os"
-	"path/filepath"
 	"sort"
-	"time"
 
+	"github.com/deref/exo/chrono"
 	"github.com/deref/exo/gensym"
 	"github.com/deref/exo/logd/api"
 	"github.com/deref/exo/logd/store"
-	"github.com/deref/exo/logd/store/badger"
-	"github.com/deref/exo/util/agent"
-	"github.com/deref/exo/util/jsonutil"
+	"github.com/deref/exo/util/errutil"
 	"github.com/deref/exo/util/mathutil"
 	"github.com/oklog/ulid/v2"
-	"golang.org/x/sync/errgroup"
 )
 
 const (
 	defaultNextLimit = 500
 	maxLimit         = 10000
 )
 
-type Config struct {
-	VarDir string
-	Debug  bool
-}
-
-func NewLogCollector(ctx context.Context, cfg *Config) *LogCollector {
-	eg, ctx := errgroup.WithContext(ctx)
-	return &LogCollector{
-		varDir:    cfg.VarDir,
-		idGen:     gensym.NewULIDGenerator(ctx),
-		debug:     cfg.Debug,
-		statePath: filepath.Join(cfg.VarDir, "logd.json"),
-		eg:        eg,
-		agent:     agent.NewAgent(300),
-	}
-}
-
 type LogCollector struct {
-	varDir string
-	idGen  *gensym.ULIDGenerator
-	debug  bool
-
-	state
-	statePath string
-
-	store   store.Store
-	eg      *errgroup.Group
-	agent   *agent.Agent
-	workers map[string]*collectorWorker
-}
-
-type state struct {
-	Logs map[string]*logState `json:"logs"`
-}
-
-type logState struct {
-	Source string `json:"source"`
-}
-
-type collectorWorker struct {
-	Worker
-	stop func()
+	Debug bool
+	IDGen *gensym.ULIDGenerator
+	Store store.Store
 }
 
 func (lc *LogCollector) debugf(format string, v ...interface{}) {
-	if lc.debug {
+	if lc.Debug {
 		fmt.Fprintln(os.Stderr, "collector", fmt.Errorf(format, v...))
 	}
 }
 
-func (lc *LogCollector) loadState() error {
-	return jsonutil.UnmarshalFile(lc.statePath, &lc.state)
-}
-
-func (lc *LogCollector) saveState() error {
-	return jsonutil.MarshalFile(lc.statePath, &lc.state)
-}
-
-func (lc *LogCollector) Run(ctx context.Context) error {
-	lc.workers = make(map[string]*collectorWorker)
-
-	if err := lc.loadState(); err != nil {
-		return fmt.Errorf("loading state: %w", err)
+func (lc *LogCollector) AddEvent(ctx context.Context, input *api.AddEventInput) (*api.AddEventOutput, error) {
+	if input.Log == "" {
```

Comment:
Do we handle pointer-type inputs? Not that there is a case for a log that is an empty string, but I'm not clear on the semantics of `*string`  vs `string` input types.

Reply:
No hard and fast rule yet. In this case, empty and nil are both invalid, so this handles both via a combo of static and dynamic means. Need more general validation logic though as we get stricter about this, I imagine.